File: CS/DXSampleTaskbarButtonService/ViewModel/MainViewModel.cs
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using DevExpress.Mvvm.DataAnnotations;
using System;
using System.Windows.Shell;

namespace DXSampleTaskbarButtonService.ViewModel {
    [POCOViewModel]
    public class MainViewModel {
        private int ThumbnailClipMargin { get; set; }
        public virtual string Description { get; set; }
        public virtual double ProgressValue { get; set; }
        [BindableProperty(OnPropertyChangedMethodName = "UpdateProgressState")]
        public virtual bool IsNormalProgressState { get; set; }
        [BindableProperty(OnPropertyChangedMethodName = "UpdateProgressState")]
        public virtual bool IsPausedProgressState { get; set; }
        [BindableProperty(OnPropertyChangedMethodName = "UpdateProgressState")]
        public virtual bool IsErrorProgressState { get; set; }
        [BindableProperty(OnPropertyChangedMethodName = "UpdateProgressState")]
        public virtual bool IsIndetermintateProgressState { get; set; }
        [BindableProperty(OnPropertyChangedMethodName = "UpdateProgressState")]
        public virtual bool IsNoneProgressState { get; set; }

        protected virtual ITaskbarButtonService TaskbarButtonService { get { return null; } }

        public MainViewModel() {
            Description = "Hello, World!";
            ProgressValue = 0.5;
            IsNormalProgressState = true;
            ThumbnailClipMargin = 0;
        }
        protected void UpdateProgressState() {
            if (TaskbarButtonService == null) return;
            if (IsNormalProgressState)
                TaskbarButtonService.ProgressState = TaskbarItemProgressState.Normal;
            else if (IsPausedProgressState)
                TaskbarButtonService.ProgressState = TaskbarItemProgressState.Paused;
            else if (IsErrorProgressState)
                TaskbarButtonService.ProgressState = TaskbarItemProgressState.Error;
            else if (IsIndetermintateProgressState)
                TaskbarButtonService.ProgressState = TaskbarItemProgressState.Indeterminate;
            else if (IsNoneProgressState)
                TaskbarButtonService.ProgressState = TaskbarItemProgressState.None;
            else
                throw new NotImplementedException();
        }

        public void IncreaseProgress() {
            if (ProgressValue < 1) {
                ProgressValue += 0.1d;
            }
        }

        public void DecreaseProgress() {
            if (ProgressValue > 0) {
                ProgressValue -= 0.1d;
            }
        }
    }
}

File: VB/DXSampleTaskbarButtonService/ViewModel/MainViewModel.vb
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations
Imports System
Imports System.Windows.Shell

Namespace DXSampleTaskbarButtonService.ViewModel

    <POCOViewModel>
    Public Class MainViewModel

        Private Property ThumbnailClipMargin As Integer

        Public Overridable Property Description As String

        Public Overridable Property ProgressValue As Double

        <BindableProperty(OnPropertyChangedMethodName:="UpdateProgressState")>
        Public Overridable Property IsNormalProgressState As Boolean

        <BindableProperty(OnPropertyChangedMethodName:="UpdateProgressState")>
        Public Overridable Property IsPausedProgressState As Boolean

        <BindableProperty(OnPropertyChangedMethodName:="UpdateProgressState")>
        Public Overridable Property IsErrorProgressState As Boolean

        <BindableProperty(OnPropertyChangedMethodName:="UpdateProgressState")>
        Public Overridable Property IsIndetermintateProgressState As Boolean

        <BindableProperty(OnPropertyChangedMethodName:="UpdateProgressState")>
        Public Overridable Property IsNoneProgressState As Boolean

        Protected Overridable ReadOnly Property TaskbarButtonService As ITaskbarButtonService
            Get
                Return Nothing
            End Get
        End Property

        Public Sub New()
            Description = "Hello, World!"
            ProgressValue = 0.5
            IsNormalProgressState = True
            ThumbnailClipMargin = 0
        End Sub

        Protected Sub UpdateProgressState()
            If TaskbarButtonService Is Nothing Then Return
            If IsNormalProgressState Then
                TaskbarButtonService.ProgressState = TaskbarItemProgressState.Normal
            ElseIf IsPausedProgressState Then
                TaskbarButtonService.ProgressState = TaskbarItemProgressState.Paused
            ElseIf IsErrorProgressState Then
                TaskbarButtonService.ProgressState = TaskbarItemProgressState.Error
            ElseIf IsIndetermintateProgressState Then
                TaskbarButtonService.ProgressState = TaskbarItemProgressState.Indeterminate
            ElseIf IsNoneProgressState Then
                TaskbarButtonService.ProgressState = TaskbarItemProgressState.None
            Else
                Throw New NotImplementedException()
            End If
        End Sub

        Public Sub IncreaseProgress()
            If ProgressValue < 1 Then
                ProgressValue += 0.1R
            End If
        End Sub

        Public Sub DecreaseProgress()
            If ProgressValue > 0 Then
                ProgressValue -= 0.1R
            End If
        End Sub
    End Class
End Namespace
